connector/spanmetricsconnector/internal/metrics/metrics.go

```diff
@@ -0,0 +1,258 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/connector/spanmetricsconnector/internal/metrics"
+
+import (
+	"sort"
+	"time"
+
+	"github.com/lightstep/go-expohisto/structure"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+type Key string
+
+type HistogramMetrics interface {
+	GetOrCreate(key Key, attributes pcommon.Map) Histogram
+	BuildMetrics(pmetric.Metric, pcommon.Timestamp, pmetric.AggregationTemporality)
+	Reset(onlyExemplars bool)
+}
+
+type Histogram interface {
+	Observe(value float64)
+	AddExemplar(traceID pcommon.TraceID, spanID pcommon.SpanID, value float64)
+}
+
+type ExplicitHistogramMetrics struct {
+	Metrics map[Key]*ExplicitHistogram
```

Comment:
There shouldn't be a need to export any of the fields from `ExplicitHistogramMetrics`, `ExponentialHistogramMetrics` (these can be set via a `New...` constructor), as well as `ExplicitHistogram` and `ExponentialHistogram` because all their accesses are made within their own package, especially now that we've moved `BuildMetrics` logic here.

Reply:
Yes, I've just started working on #18502 and un-exported all those fields there and added `New` factory methods. I'll do that in this PR then.

Reply:
Pushed.